contracts/token/ERC721/ERC721.sol

```diff
@@ -142,19 +143,19 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Enumerable
 
         string memory _tokenURI = _tokenURIs[tokenId];
 
-        // Even if there is a base URI, it is only appended to non-empty token-specific URIs
-        if (bytes(_tokenURI).length == 0) {
-            return "";
+        // If there is no token-specific URI but a base URI, append the token ID to the base URI.
+        // Otherwise, append the token URI to the base URI.
+        // In both cases, abi.encodePacked is being used to concatenate strings.
+        if (bytes(_tokenURI).length == 0 && bytes(_baseURI).length > 0) {
```

Comment:
Could this be written as follows? I feel this way we emphasize the fact that setting a base URI is an opt-in for the automatic token URI behavior.

```solidity
// If there is no base URI, return the token URI
if (bytes(_baseURI).length == 0) {
    return _tokenURI;
// Otherwise concatenate the baseURI and tokenURI
} else if (bytes(_tokenURI).length > 0) {
    return string(abi.encodePacked(_baseURI, _tokenURI)); 
// If there is no tokenURI, concatenate the tokenID to the base URI
} else {
    return string(abi.encodePacked(_baseURI, Strings.fromUint256(tokenId))); 
}
```

Reply:
Sure, I'll update the PR tomorrow. And actually, we don't need to use `else` at all, as `return` omits the rest anyhow. Or do you prefer having explicit `else` branches?

Reply:
I have no particular preference over early returns and exhaustive if-else blocks :shrug: The big thing is we're now making it explicit that there's 3 different ways this can work, which was not clear on the original version.